docs/standard/native-interop/best-practices.md

```diff
@@ -251,6 +251,56 @@ finally
 }
 ```
 
+## Windows and non-Windows data type considerations
+
+There are types in the C language that have latitude in how they are defined. When writing cross-platform interop, cases can arise where platforms differ and can cause issues if not considered.
+
+### C `long`
+C `long` and C# `long` are not the same types. Using C# `long` to interop with C `long` is almost never correct.
+The `long` keyword in C is defined to have ["at least 32"](https://en.cppreference.com/w/c/language/arithmetic_types) bits. This means there is a minimum required but platforms are not limited to that. The table below illustrates the differences in provided bits for the C `long` data type between platforms.
+
+|             | 32-bit | 64-bit |
+|:------------|:-------|:-------|
+| Windows     | 32     | 32     |
+| macOS/\*nix | 32     | 64     |
+
+These differences make authoring cross-platform P/Invokes difficult when the native function is defined to use `long` on all platforms. When targeting .NET 5 and prior .NET versions, you should declare separate Windows and non-Windows signatures to handle the problem.
+
+```csharp
+static readonly bool IsWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+
+// Cross platform C function
+// void Function(long a);
+
+[DllImport("NativeLib", EntryPoint = "Function")]
+extern static void FunctionWindows(int a);
+
+[DllImport("NativeLib", EntryPoint = "Function")]
+extern static void FunctionUnix(nint a);
+
+// Usage
+if (IsWindows)
+{
+    FunctionWindows(a);
+}
+else
+{
+    FunctionUnix(a);
+}
+```
+
+In .NET 6 a new type has been introduced to help ameliorate this issue - [`CLong` and `CULong`](https://github.com/dotnet/runtime/issues/13788). In .NET 6 the following is now possible. The example is for `CLong`, but you can use `CULong` to abstract C's `unsigned long` in a similar way.
+
+```csharp
+// Cross platform C function
+// void Function(long a);
+[DllImport("NativeLib")]
+extern static void Function(CLong a);
+    
+// Usage
+Function(a);
+```
+
 ## Structs
```

Comment:
The doc jumps back and forth between general guidelines and very specific Windows stuff. I think overall flow of this document can be improved. It can be done in a different PR. 

General guidelines - strings, structs, fixed buffers, etc.
Cross-platform data type considerations
Platform-specific - Common Windows data types, Formerly built-in supported types

It may be even fine to split this into multiple docs. I see that we have [Writing cross platform P/Invoke code](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/cross-platform). Should the "Cross-platform data type considerations" topic be there?

Reply:
Yes. This is part of the exercise being done with the interop-doc - what is a logical flow for learning about interop and best to walk users through writing interop solutions.